src/components/avatar-audio-source.js

```diff
@@ -0,0 +1,115 @@
+  return audioEl;
+}
+
+function getOrCreateAudioListenerForScene(scene) {
+  if (!scene.audioListener) {
+    // Would prefer not to have to do this, and am not sure if it's required,
+    // so adding a warning for now and hoping to remove this later.
+    console.warn("Scene did not have an audio listener. Creating one now.");
+    scene.audioListener = new THREE.AudioListener();
+    scene.camera && scene.camera.add(scene.audioListener);
+    scene.addEventListener("camera-set-active", function(evt) {
+      evt.detail.cameraEl.getObject3D("camera").add(scene.audioListener);
+    });
+  }
+  return scene.audioListener;
+}
+
+async function getMediaStream(el) {
+  const networkedEl = await NAF.utils.getNetworkedEntity(el).catch(e => {
+    console.error(INFO_INIT_FAILED, INFO_NO_NETWORKED_EL, e);
+  });
+  if (!networkedEl) {
+    return null;
+  }
+  const ownerId = networkedEl.components.networked.data.owner;
+  if (!ownerId) {
+    console.error(INFO_INIT_FAILED, INFO_NO_OWNER);
+    return null;
+  }
+  const stream = await NAF.connection.adapter.getMediaStream(ownerId).catch(e => {
+    console.error(INFO_INIT_FAILED, `Error getting media stream for ${ownerId}`, e);
+  });
+  if (!stream) {
+    return null;
+  }
+  return stream;
+}
+
+AFRAME.registerComponent("avatar-audio-source", {
+  schema: {
+    positional: { default: true },
+    distanceModel: {
+      default: "inverse",
+      oneOf: ["linear", "inverse", "exponential"]
+    },
+    maxDistance: { default: 10000 },
+    refDistance: { default: 1 },
+    rolloffFactor: { default: 1 }
+  },
+
+  init: async function() {
+    this.el.sceneEl.systems["hubs-systems"].audioSettingsSystem.registerAvatarAudioSource(this);
+    const stream = await getMediaStream(this.el);
+    const isRemoved = !this.el.parentNode;
+    if (!stream || isRemoved) return;
+
+    const audioListener = getOrCreateAudioListenerForScene(this.el.sceneEl);
+    const audio = this.data.positional ? new THREE.PositionalAudio(audioListener) : new THREE.Audio(audioListener);
+    if (this.data.positional) {
+      audio.setDistanceModel(this.data.distanceModel);
+      audio.setMaxDistance(this.data.maxDistance);
+      audio.setRefDistance(this.data.refDistance);
+      audio.setRolloffFactor(this.data.rolloffFactor);
+    }
+
+    if (SHOULD_CREATE_SILENT_AUDIO_ELS) {
+      createSilentAudioEl(stream); // TODO: Do the audio els need to get cleaned up?
+    }
+
+    const mediaStreamSource = audio.context.createMediaStreamSource(stream);
+    audio.setNodeSource(mediaStreamSource);
+    this.el.emit("sound-source-set", { soundSource: mediaStreamSource });
+    this.el.setObject3D(this.attrName, audio);
+  },
+
+  update(oldData) {
+    const audio = this.el.getObject3D(this.attrName);
+    if (!audio) return;
+
+    const shouldRecreateAudio = oldData.positional !== this.data.positional;
+    if (shouldRecreateAudio) {
+      audio.disconnect();
+      this.el.removeObject3D(this.attrName);
```

Comment:
The overlap between this and what `remove` does should also maybe be abstracted

Reply:
:+1: